crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -1060,86 +1060,91 @@ impl<'a, 'o> TimelineEventHandler<'a, 'o> {
-                    //       old and new item?
+                        // TODO: Check whether anything is different about the
+                        //       old and new item?
 
-                    transfer_details(&mut item, &old_item);
+                        transfer_details(&mut item, &old_item);
 
-                    let old_item_id = old_item.internal_id;
+                        let old_item_id = old_item.internal_id;
 
-                    if idx == self.items.len() - 1 {
-                        // If the old item is the last one and no day divider
-                        // changes need to happen, replace and return early.
-                        trace!(idx, "Replacing existing event");
-                        self.items.set(idx, TimelineItem::new(item, old_item_id.to_owned()));
-                        return;
-                    }
+                        if idx == self.items.len() - 1 {
+                            // If the old item is the last one and no day divider
+                            // changes need to happen, replace and return early.
+                            trace!(idx, "Replacing existing event");
+                            self.items.set(idx, TimelineItem::new(item, old_item_id.to_owned()));
+                            return;
+                        }
 
-                    // In more complex cases, remove the item before re-adding the item.
-                    trace!("Removing local echo or duplicate timeline item");
-                    removed_event_item_id = Some(self.items.remove(idx).internal_id.clone());
+                        // In more complex cases, remove the item before re-adding the item.
+                        trace!("Removing local echo or duplicate timeline item");
 
-                    // no return here, below code for adding a new event
-                    // will run to re-add the removed item
-                }
+                        // no return here, below code for adding a new event
+                        // will run to re-add the removed item
 
-                // Local echoes that are pending should stick to the bottom,
-                // find the latest event that isn't that.
-                let latest_event_idx = self
-                    .items
-                    .iter()
-                    .enumerate()
-                    .rev()
-                    .find_map(|(idx, item)| (!item.as_event()?.is_local_echo()).then_some(idx));
+                        Some(self.items.remove(idx).internal_id.clone())
+                    } else {
+                        None
+                    }
+                };
 
                 // Insert the next item after the latest event item that's not a
                 // pending local echo, or at the start if there is no such item.
-                let insert_idx = latest_event_idx.map_or(0, |idx| idx + 1);
+                let (insert_index, must_push_front) = match position {
```

Comment:
can we keep the shorter `insert_idx` name, please?

Reply:
It becomes:

```rust
if !must_push_front && (insert_index == self.items.len()) {
    self.items.push_back(new_item);
} else if must_push_front || insert_index == 0 {
    self.items.push_front(new_item);
} else {
    self.items.insert(insert_index, new_item);
}
```

which can be rewritten to:

```rust
if !must_push_front {
    if insert_index == self.items.len() {
        self.items.push_back(new_item);
    } else {
        self.items.insert(insert_index, new_item);
    }
} else {
    self.items.push_front(new_item);
}
```

for a bit more clarity.

Reply:
Done.